File: Power Triangle.py
# -*- coding: utf-8 -*-
# ----------------------------------------------------------------------------
# Power Triangle.py
#
# Created on: 2017-04-13
# Stephan Garland
# stephan.marc.garland@gmail.com
#
# Given two inputs (kW, kVA, kVAR, PF, or theta), calculates the others
# and graphs the resulting triangle as a PNG (requires matplotlib)



import math as m
try:
    import tkinter as tk  # Python 3.x
    from tkinter import messagebox as tkMsg
except ImportError:  # Python 2.x
    import Tkinter as tk
    import tkMessageBox as tkMsg
import matplotlib.pyplot as plt


fields = 'kW', 'kVA', 'kVAR', 'PF', 'Theta'
labels = {'kW':'kW (Real)', 'kVA':'kVA (Apparent)', 'kVAR':'kVAR (Reactive)',\
          'PF':'PF (0.00-1.00)', 'Theta':'Theta (Degrees)'}

def fetch(entries):
    inputs = {}
    for entry in entries:
        field = entry[0]
        text = entry[1].get() # User input for each textbox
        if text: # True if field isn't blank
            inputs[field] = float(text)
    if inputs.get('Theta') is not None: # If Theta isn't given, don't convert
        if inputs.get('Theta') > 0 and inputs.get('Theta') <= 90:
            inputs['Theta'] = m.radians(inputs['Theta']) # Use radians internally
        else:
            tkMsg.showerror("Error","Theta must be 0 < x < 90")
            quitProg()
    if inputs.get('kVA',0) <= max(inputs.get('kW',0),inputs.get('kVAR',0))\
        and inputs.get('kVA') is not None:
            tkMsg.showerror("Error","kVA must be > any other side")
            quitProg()
    if inputs.get('PF') is not None:
        if inputs.get('PF') < 0 or inputs.get('PF') > 1:
            pfError = tkMsg.askyesno("PF out of range", "You entered {0} for PF\
            \nDid you mean {1}?".format(inputs['PF'],inputs['PF']/100))
            if pfError == 1:
                inputs['PF'] = inputs['PF']/100
            if pfError == 0:
                tkMsg.showerror("Error","PF must be 0 < x < 1")
                quitProg() 
    global calcInputs
    calcInputs = inputs
    root.quit()
        


def makeform(root, fields):
    entries = []
    for field in fields:
        row = tk.Frame(root)
        lab = tk.Label(row, width=25, text=labels[field], anchor='w')
        ent = tk.Entry(row)
        row.pack(side=tk.TOP, fill=tk.X, padx=5, pady=5)
        lab.pack(side=tk.LEFT)
        ent.pack(side=tk.RIGHT, expand=tk.YES, fill=tk.X)
        entries.append((field, ent))
    return entries

    
    
def resultsWindow(fields):
    root.withdraw()
    top = tk.Toplevel()
    b3 = tk.Button(top, text = 'Quit', command=quitProg)
    b3.pack(side=tk.RIGHT, padx=5, pady=5)
    for field in fields:
        row = tk.Frame(top)
        lab = tk.Label(row, width=25, text=field, anchor='w')
        row.pack(side=tk.TOP, fill=tk.X, padx=5, pady=5)
        lab.pack(side=tk.LEFT)
    

    
def returnResults(kW, kVAR, kVA, PF, Theta):
    d = locals()
    roundedOut = {}
    for k,v in d.items():
        if k == 'Theta':
            # Cast theta to a string to allow concatenating with degree symbol
            # Also convert theta to degrees; all internal calcs are in radians
            roundedOut['Theta'] = str(round(m.degrees(v),3)) + "\xb0"
        else:
            roundedOut[k] = round(v,3)
    root.quit()
    resultsWindow(roundedOut.items())
    drawViz(kW, kVAR, kVA, m.degrees(Theta))
    
    
    
def from_KW_KVA(kW, kVA):
    Theta = m.acos(kW / kVA)
    kVAR = m.sin(Theta) * kVA
    PF = m.cos(Theta)
    returnResults(kW, kVAR, kVA, PF, Theta)


    
def from_KW_KVAR(kW, kVAR):
    Theta = m.atan(kVAR / kW)
    PF = m.cos(Theta)
    kVA = kW / PF
    returnResults(kW, kVAR, kVA, PF, Theta)
    
    
    
def from_KW_PF(kW, PF):
    Theta = m.acos(PF)
    kVAR = m.tan(Theta) * kW
    kVA = kW / PF
    returnResults(kW, kVAR, kVA, PF, Theta)
        
        
        
def from_KW_THETA(kW, Theta):
    kVAR = m.tan(Theta) * kW
    PF = m.cos(Theta)
    kVA = kW / PF
    returnResults(kW, kVAR, kVA, PF, Theta)
    
    
def from_KVA_KVAR(kVA, kVAR):
    Theta = (m.asin(kVAR/kVA))
    PF = m.cos(Theta)
    kW = kVA * PF
    returnResults(kW, kVAR, kVA, PF, Theta)
    
    
    
def from_KVA_PF(kVA, PF):
    Theta = m.acos(PF)
    kW = kVA * PF
    kVAR = m.tan(Theta)*kW
    returnResults(kW, kVAR, kVA, PF, Theta)



def from_KVA_THETA(kVA, Theta):
    PF = m.cos(Theta)
    kW = kVA * PF
    kVAR = m.tan(Theta)*kW
    returnResults(kW, kVAR, kVA, PF, Theta)



def from_KVAR_PF(kVAR, PF):
    Theta = m.acos(PF)
    kVA = kVAR / m.sin(Theta)
    kW = kVA * PF
    returnResults(kW, kVAR, kVA, PF, Theta)



def from_KVAR_THETA(kVAR, Theta):
    PF = m.cos(Theta)
    kVA = kVAR / m.sin(Theta)
    kW = kVA * PF
    returnResults(kW, kVAR, kVA, PF, Theta)
    
    

def catchInput(PF, Theta):
    tkMsg.showwarning("Inadequate Data", "Power Factor is cos(theta).\nPlease use other values.")
    quitProg()
    
    
    
def getInput():
    root.deiconify()
    root.title("Power Triangle Calculator")
    ents = makeform(root, fields)
    root.bind('<Return>', (lambda event, e=ents: fetch(e)))
    b1 = tk.Button(root, text = 'Calculate', command=(lambda e=ents: fetch(e)))
    b1.pack(side=tk.LEFT, padx=5, pady=5)
    b2 = tk.Button(root, text = 'Quit', command=quitProg)
    b2.pack(side=tk.LEFT, padx=5, pady=5)
    root.mainloop()

    
    
def drawViz(kW, kVAR, kVA, Theta):
    p = [0, 0] # x = 0, y = 0 (real power base)
    q = [kW, 0] # x = kW, y = 0 (real power scalar / reactive power base)
    s = [kW, kVAR] # x = kW, y = kVAR (apparent power top)
    slope = kVAR/kW
    points = [p, q, s]
    plt.axes()
    polygon = plt.Polygon(points)
    plt.gca().add_patch(polygon)
    plt.axis('scaled')
    # All xy values were established with trial and error, change as desired
    plt.annotate('{0} kW'.format(round(kW,2)), xy=(0.5,-0.1), xycoords='axes fraction')
    plt.annotate('{0} kVAR'.format(round(kVAR,2)), xy=(1.05,0.5), xycoords='axes fraction')
    plt.annotate('{0} kVA'.format(round(kVA,2)), xy=((0.25*kW),\
        (0.5*(slope*kW)+(0.05*kW))), xycoords='data')
    #$\\x$ is LaTeX encoding for matplotlib
    plt.annotate(u'$\\theta$ {0} \xb0'.format(round(Theta,3)),\
        xy=((0.05*kW),(0.02*kVAR)), xycoords='data')
    plt.show()
 

 
def restartProg():
    #root.quit()
    main()
    
    

def quitProg():
    raise SystemExit
    
    

def main():
    root.withdraw()

    getInput()

    given_keys = []

    for key in sorted(fields):
        if key in calcInputs:
            given_keys.append(key)

    calc_key = '_'.join(given_keys)

    dispatch = {

            'kW_kVA' : from_KW_KVA,
            'kVA_kW' : from_KW_KVA,
            'kW_kVAR' : from_KW_KVAR,
            'kVAR_kW' : from_KW_KVAR,
            'kW_PF' : from_KW_PF,
            'PF_kW' : from_KW_PF,
            'kW_Theta' : from_KW_THETA,
            'Theta_kW' : from_KW_THETA,
            'kVA_kVAR' : from_KVA_KVAR,
            'kVAR_kVA' : from_KVA_KVAR,
            'kVA_PF' : from_KVA_PF,
            'PF_kVA' : from_KVA_PF,
            'kVA_Theta' : from_KVA_THETA,
            'Theta_kVA' : from_KVA_THETA,
            'kVAR_PF' : from_KVAR_PF,
            'PF_kVAR' : from_KVAR_PF,
            'kVAR_Theta' : from_KVAR_THETA,
            'Theta_kVAR' : from_KVAR_THETA,
            'PF_Theta' : catchInput,
            'Theta_PF' : catchInput
            
                } # PF_Theta doesn't work as PF is cos(theta)
                  # Couldn't find a better way to sort, so foo_bar | bar_foo works
                
    calc_function = dispatch[calc_key] # e.g. calc_function becomes from_KW_KVA
    calc_function(**calcInputs) # e.g. from_KW_KVA is called with calcInputs as kwargs

    
root = tk.Tk()
root.withdraw()    
tkMsg.showinfo("Usage", "Please enter only two values.\nTheta and Power Factor cannot be used together.")
main()

File: Power Triangle.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;



namespace Power_Triangle
{
    public partial class Form1 : Form
    {
        double kw = 0;
        double kvar = 0;
        double kva = 0;
        double pf = 0;
        double theta = 0;

        public Form1()
        {
            InitializeComponent();
        }



        private void textBox5_TextChanged(object sender, EventArgs e)
        {

        }


        private void textBox4_TextChanged(object sender, EventArgs e)
        {

        }



        private void textBox3_TextChanged(object sender, EventArgs e)
        {

        }



        private void textBox2_TextChanged(object sender, EventArgs e)
        {

        }



        private void textBox1_TextChanged(object sender, EventArgs e)
        {

        }



        private void button1_Click(object sender, EventArgs e)
        {
            Dispatch();
        }



        private void button3_Click(object sender, EventArgs e)
        {
            Close();
        }



        private void button2_Click(object sender, EventArgs e)
        {
            ResetForm();
        }



        private void ResetForm()
        {
            foreach (var c in this.Controls)
            {
                if (c is TextBox)
                {
                    ((TextBox)c).Text = String.Empty;
                }
            }
        }



        private Dictionary<string, double> GetInput()
        {
            // Make dict with text containing input, label relevant portion of form label
            // Keys are the labels on the form, with extraneous info removed
            // Values are the raw contents of the text boxes, null or otherwise

            var label_var = new Dictionary<string, string>();
            // Split twice to extract title, trim as there is leading and trailing whitespace
            string local_label_1 = label1.ToString().Split('(')[0].Split(':')[1].Trim();
            string local_label_2 = label2.ToString().Split('(')[0].Split(':')[1].Trim();
            string local_label_3 = label3.ToString().Split('(')[0].Split(':')[1].Trim();
            string local_label_4 = label4.ToString().Split('(')[0].Split(':')[1].Trim();
            string local_label_5 = label5.ToString().Split('(')[0].Split(':')[1].Trim();
            label_var[local_label_1] = textBox1.Text;
            label_var[local_label_2] = textBox2.Text;
            label_var[local_label_3] = textBox3.Text;
            label_var[local_label_4] = textBox4.Text;
            label_var[local_label_5] = textBox5.Text;


            // Make dict to hold only the non-null values
            var calc_vars = new Dictionary<string, double>();

            // Try to convert all values to doubles, giving them names used later
            // They're in a try block so the blanks can get caught
            // In the catch block, the foreach puts the two non-nulls into calc_vars dict
            try
            {
                kw = Convert.ToDouble(label_var[local_label_1]);
                kva = Convert.ToDouble(label_var[local_label_2]);
                kvar = Convert.ToDouble(label_var[local_label_3]);
                pf = Convert.ToDouble(label_var[local_label_4]);
                theta = Convert.ToDouble(label_var[local_label_5]);
            }

            catch (FormatException)
            {
                foreach (var pair in label_var)
                {
                    if (!String.IsNullOrEmpty(pair.Value))
                    {
                        calc_vars.Add(pair.Key, Convert.ToDouble(pair.Value));
                    }
                }
            }
            return calc_vars;
        }



        private void Dispatch()
        {
            // dispatch_input == calc_vars, which is a Dictionary (string/double)
            var dispatch_input = GetInput();
            var dispatch_list = new List<string>();

            foreach (var dispatch_var in dispatch_input)
            {
                dispatch_list.Add(dispatch_var.Key);
            }

            string dispatch_key = String.Join("_", dispatch_list.ToArray());

            switch(dispatch_key)
            {
                case "kW_kVA":
                    FromKwKva(dispatch_input["kW"], dispatch_input["kVA"]);
                    break;
                case "kW_kVAR":
                    FromKwKvar(dispatch_input["kW"], dispatch_input["kVAR"]);
                    break;
                case "kW_PF":
                    FromKwPf(dispatch_input["kW"], dispatch_input["PF"]);
                    break;
                case "kW_Theta":
                    FromKwTheta(dispatch_input["kW"], dispatch_input["Theta"]);
                    break;
                case "kVA_kVAR":
                    FromKvaKvar(dispatch_input["kVA"], dispatch_input["kVAR"]);
                    break;
                case "kVA_PF":
                    FromKvaPf(dispatch_input["kVA"], dispatch_input["PF"]);
                    break;
                case "kVA_Theta":
                    FromKvaTheta(dispatch_input["kVA"], dispatch_input["Theta"]);
                    break;
                case "kVAR_PF":
                    FromKvarPf(dispatch_input["kVAR"], dispatch_input["PF"]);
                    break;
                case "kVAR_Theta":
                    FromKvarTheta(dispatch_input["kVAR"], dispatch_input["Theta"]);
                    break;
                case "PF_Theta":
                    CatchInput();
                    break;
                default:
                    MessageBox.Show("Error");
                    break;
            }
        }

        /*
        private Dictionary<string, Action<string>> dispatch_dict =
            new Dictionary<string, Action<string>>
            {

                { "kW_kVA" , FromKwKva },
                { "kVA_kW" , "FromKwKva" },
                { "kW_kVAR" , "FromKwKvar" },
                { "kVAR_kW" , "FromKwKvar" },
                { "kW_PF" , "FromKwPf" },
                { "PF_kW" , "FromKwPf" },
                { "kW_Theta" , "FromKwTheta" },
                { "Theta_kW" , "FromKwTheta" },
                { "kVA_kVAR" , "FromKvaKvar" },
                { "kVAR_kVA" , "FromKvaKvar" },
                { "kVA_PF" , "FromKvaPf" },
                { "PF_kVA" , "FromKvaPf" },
                { "kVA_Theta" , "FromKvaTheta" },
                { "Theta_kVA" , "FromKvaTheta" },
                { "kVAR_PF" , "FromKvarPf" },
                { "PF_kVAR" , "FromKvarPf" },
                { "kVAR_Theta" , "FromKvarTheta" },
                { "Theta_kVAR" , "FromKvarTheta" },
                { "PF_Theta" , "CatchInput" },
                { "Theta_PF" , "CatchInput" }

            };
            */


        private double RadianToDegree(double angle_in_radians)
        {
            return angle_in_radians * (180 / Math.PI);
        }



        private double DegreeToRadian(double angle_in_degrees)
        {
            return angle_in_degrees * (Math.PI / 180);
        }



        private void FromKwKva(double kw, double kva)
        {
            theta = Math.Acos(kw / kva);
            kvar = Math.Sin(theta) * kva;
            pf = Math.Cos(theta);
            ReturnResults(kw, kvar, kva, pf, theta);
        }



        private void FromKwKvar(double kw, double kvar)
        {
            theta = Math.Atan(kvar / kw);
            pf = Math.Cos(theta);
            kva = kw / pf;
            ReturnResults(kw, kvar, kva, pf, theta);
        }



        private void FromKwPf(double kw, double pf)
        {
            theta = Math.Acos(pf);
            kvar = Math.Tan(theta) * kw;
            kva = kw / pf;
            ReturnResults(kw, kvar, kva, pf, theta);
        }



        private void FromKwTheta(double kw, double theta)
        {
            theta = DegreeToRadian(theta);
            kvar = Math.Tan(theta) * kw;
            pf = Math.Cos(theta);
            kva = kw / pf;
            ReturnResults(kw, kvar, kva, pf, theta);
        }



        private void FromKvaKvar(double kva, double kvar)
        {
            theta = Math.Asin(kvar / kva);
            pf = Math.Cos(theta);
            kw = kva * pf;
            ReturnResults(kw, kvar, kva, pf, theta);

        }



        private void FromKvaPf(double kva, double pf)
        {
            theta = Math.Acos(pf);
            kw = kva * pf;
            kvar = Math.Tan(theta) * kw;
            ReturnResults(kw, kvar, kva, pf, theta);

        }



        private void FromKvaTheta(double kva, double theta)
        {
            theta = DegreeToRadian(theta);
            pf = Math.Cos(theta);
            kw = kva * pf;
            kvar = Math.Tan(theta) * kw;
            ReturnResults(kw, kvar, kva, pf, theta);
        }



        private void FromKvarPf(double kvar, double pf)
        {
            theta = Math.Acos(pf);
            kva = kvar / Math.Sin(theta);
            kw = kva * pf;
            ReturnResults(kw, kvar, kva, pf, theta);
        }



        private void FromKvarTheta(double kvar, double theta)
        {
            theta = DegreeToRadian(theta);
            pf = Math.Cos(theta);
            kva = kvar / Math.Sin(theta);
            kw = kva * pf;
            ReturnResults(kw, kvar, kva, pf, theta);
        }
        


        private void CatchInput()
        {
            MessageBox.Show("Power Factor is cos(theta).\nPlease use other values.", "Inadequate Data");
        }



        private void ReturnResults(double kw, double kvar, double kva, double pf, double theta)
        {
            theta = RadianToDegree(theta); // Everything is done internally in radians

            string kw_disp = Convert.ToString(Math.Round(kw, 3));
            string kva_disp = Convert.ToString(Math.Round(kva, 3));
            string kvar_disp = Convert.ToString(Math.Round(kvar, 3));
            string pf_disp = Convert.ToString(Math.Round(pf, 3));
            string theta_disp = Convert.ToString(Math.Round(theta, 3));

            // This foreach will hit the non-text boxes, but they are ignored
            foreach (Control text_box_edit in this.Controls)
            {
                switch (text_box_edit.Name)
                {
                    case "textBox1":
                        text_box_edit.Text = kw_disp;
                        break;
                    case "textBox2":
                        text_box_edit.Text = kva_disp;
                        break;
                    case "textBox3":
                        text_box_edit.Text = kvar_disp;
                        break;
                    case "textBox4":
                        text_box_edit.Text = pf_disp;
                        break;
                    case "textBox5":
                        text_box_edit.Text = theta_disp;
                        break;
                }
            }
        }
    }
}

// Allows ToDegrees attribute extension to function in 3.0 framework.
namespace System.Runtime.CompilerServices
{
    class ExtensionAttribute : Attribute
    {

    }
}
